### ISSUE TITLE
Fails with Math::BigInt 1.999801

### ISSUE THREAD

**Issue by andk** [DCIT/perl-CryptX#26]
As per subject. Diagnostics:

```
#   Failed test '$x = Math::BigInt->new("+0"); $y = Math::BigInt->new("abc"); Math::BigInt::blcm($x, $y);'
#   at t/mbi_ltm/bigintpm.inc line 202.
#          got: undef
#     expected: 'NaN'

#   Failed test '$x = Math::BigInt->new("+0"); $y = Math::BigInt->new("+0"); Math::BigInt::blcm($x, $y);'
#   at t/mbi_ltm/bigintpm.inc line 202.
#          got: undef
#     expected: 'NaN'

#   Failed test '$x = Math::BigInt->new("+1"); $y = Math::BigInt->new("+0"); Math::BigInt::blcm($x, $y);'
#   at t/mbi_ltm/bigintpm.inc line 202.
#          got: undef
#     expected: '0'

#   Failed test '$x = Math::BigInt->new("+0"); $y = Math::BigInt->new("+1"); Math::BigInt::blcm($x, $y);'
#   at t/mbi_ltm/bigintpm.inc line 202.
#          got: undef
#     expected: '0'

#   Failed test '$x = Math::BigInt->new("+27"); $y = Math::BigInt->new("+90"); Math::BigInt::blcm($x, $y);'
#   at t/mbi_ltm/bigintpm.inc line 202.
#          got: undef
#     expected: '270'

#   Failed test '$x = Math::BigInt->new("+1034"); $y = Math::BigInt->new("+804"); Math::BigInt::blcm($x, $y);'
#   at t/mbi_ltm/bigintpm.inc line 202.
#          got: undef
#     expected: '415668'
# Looks like you planned 3730 tests but ran 3724.
# Looks like you failed 6 tests of 3724 run.
t/mbi_ltm_bigintpm.t ................ 
Dubious, test returned 6 (wstat 1536, 0x600)
Failed 12/3730 subtests 
        (less 50 skipped subtests: 3668 okay)
```

Downgrading PJACKLAM/Math-BigInt-1.999727.tar.gz does as a workaround.

HTH&&Thanks,

**Comment by karel-m** [comment]
cc @pjacklam - any hints, Peter?

**Comment by pjacklam** [comment]
To avoid all the code duplication in the Math::BigInt backend libraries, I wrote a parent class called Math::BigInt::Lib that all backend libraries inherit from. In addition, I moved stuff from Math::BigInt into Math::BigInt::Lib, where it really belonged. Because of inheritance, this was no problem for the other backend libraries (Calc, FastCalc, GMP, and Pari), but I forgot about Math::BigInt::LTM.

The problems should be gone if you specify that Math::BigInt::LTM is a subclass of Math::BigInt::Lib.

**Comment by karel-m** [comment]
Thanks, after adding:

```perl
use Math::BigInt::Lib '1.999801';
our @ISA = qw< Math::BigInt::Lib >;
```

there remains only one failing test:

```
#   Failed test '$x = Math::BigInt->new("+0"); $y = Math::BigInt->new("+0"); Math::BigInt::blcm($x, $y);'
#   at t/mbi_ltm/bigintpm.inc line 202.
#          got: '0'
#     expected: 'NaN'
```

which seems like I should simply adjust the expected result to `0` instead of `NaN`.

However, I am a bit uncomfortable with forcing CryptX to be dependent on `Math::BigInt` 1.999801 as it will cause troubles in our codebase at $work. I have to think it over.

**Comment by pjacklam** [comment]
If you copy the _lcm() method from Math::BigInt::Lib directly into Math::BigInt::LTM, all tests except one will pass. As for the remaining failed test, I noticed that LCM(0,0) by convention is 0, not NaN, so I changed the behaviour in Math::BigInt accordingly.

**Comment by karel-m** [comment]
Great, copying `_lcm()` did the trick.

I have tried to keep in sync all Math::BigInt::LTM's methods with Math::BigInt::Lib. But I have run into a trouble after I have replaced LTM's old `_log_int` (some time ago stolen from Math::BigInt::GMP) with a new one form Math::BigInt::Lib.

The following test:
```perl
$x = $C->_new("81");
my ($r, $exact) = $C->_log_int($x, $C->_new("3"));
ok($C->_str($r), '4');
ok($C->_str($x), '4');  # <<<<<<<<<<<<<<<< FAIL
ok($exact, 1);
```

started to fail. But frankly speaking I have no idea why `$x` is expected at this point to have the same value as the result `$r`.
